src/main/scala/mesosphere/marathon/core/task/termination/impl/KillStreamWatcher.scala

```diff
@@ -1,101 +1,112 @@
 
-    // eagerComplete allows us to toss the right materialized cancellable value
-    val liveInstanceIds = killedViaInstanceChanged.merge(killedViaUnknownInstanceTerminated, eagerComplete = true)
+    instanceUpdates
+      .flatMapConcat {
+        case (snapshot, updates) =>
+          val pendingInstanceIds: Set[Instance.Id] = snapshot.instances.iterator
+            .filter { i => instancesMap.contains(i.instanceId) }
+            .filterNot { i => predicate(i, instancesMap(i.instanceId)) }
+            .map(_.instanceId)
+            .to[Set]
 
-    Source(alreadyConsideredTerminal)
-      .mergeMat(liveInstanceIds)(Keep.right)
+          updates
+            .filter { change => instancesMap.contains(change.id) }
+            .scan(pendingInstanceIds) {
+              case (remaining, deleted: InstanceDeleted) =>
+                remaining - deleted.id
+              case (remaining, InstanceUpdated(instance, _, _)) if predicate(instance, instancesMap.getOrElse(instance.instanceId, Set.empty)) =>
+                remaining - instance.instanceId
+              case (remaining, _) =>
+                remaining
+            }
+            .takeWhile(_.nonEmpty, inclusive = true)
+      }
   }
 
-  private val singleDone = List(Done)
-  private[impl] def killedInstanceFlow(instanceIds: Iterable[Instance.Id]): Flow[Instance.Id, Done, NotUsed] = {
-
-    val instanceIdsSet = instanceIds.toSet
-    if (instanceIdsSet.isEmpty) {
-      logger.info("Asked to watch no instances. Completing immediately.")
-      EnrichedFlow.ignore.prepend(Source.single(Done)).take(1)
-    } else {
-      Flow[Instance.Id].
-        filter(instanceIdsSet).
-        statefulMapConcat { () =>
-          var pendingInstanceIds = instanceIdsSet
-          // this is used for logging to help prevent polluting the logs
-          val name = s"kill-watcher-${instanceIds.headOption.fold("")(i => i.safeRunSpecId)}-${UUID.randomUUID()}"
-
-          { (id: Instance.Id) =>
-            pendingInstanceIds -= id
-            logger.debug(s"Received terminal update for ${id}")
-            if (pendingInstanceIds.isEmpty) {
-              logger.info(s"${name} done watching; all watched instances were killed")
-              singleDone
-            } else {
-              logger.info(s"${name} still waiting for ${pendingInstanceIds.size} instances to be killed")
-              Nil
-            }
-          }
-        }.
-        take(1)
-    }
+  /**
+    * Wait until either the instances are expunged, the tasks associated with the instances are terminal, or have been
+    * replaced with new tasks.
+    */
+  def watchForKilledInstances(instanceUpdates: InstanceTracker.InstanceUpdates, instances: Iterable[Instance])(implicit materializer: Materializer): Future[Done] = {
+    KillStreamWatcher.
+      emitPendingTerminal(instanceUpdates, instances, whenTerminalOrReplaced)
+      .takeWhile { pendingTerminalInstances => pendingTerminalInstances.nonEmpty }
```

Comment:
Aren't we checking `nonEmpty` in line 80?

Reply:
Great point